File: packages/anki/connect.py
"""AnkiConnect HTTP client for async communication with Anki.

Provides a unified async client for all AnkiConnect API operations
including note CRUD, deck management, tags, and model queries.

Usage:
    async with AnkiConnectClient() as client:
        if await client.ping():
            decks = await client.deck_names()
"""

from __future__ import annotations

from typing import Any, Final, cast

import httpx
import structlog

from packages.common.exceptions import AnkiConnectError

logger = structlog.get_logger()

ANKI_CONNECT_URL: Final[str] = "http://localhost:8765"
ANKI_CONNECT_VERSION: Final[int] = 6
DEFAULT_TIMEOUT: Final[float] = 30.0


class AnkiConnectClient:
    """Async client for AnkiConnect API.

    Can be used as an async context manager or manually managed:

        async with AnkiConnectClient() as client:
            await client.ping()

        # or
        client = AnkiConnectClient()
        try:
            await client.ping()
        finally:
            await client.close()
    """

    def __init__(
        self,
        url: str = ANKI_CONNECT_URL,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        self.url = url
        self.timeout = timeout
        self._client: httpx.AsyncClient | None = None

    async def __aenter__(self) -> AnkiConnectClient:
        self._client = httpx.AsyncClient(timeout=self.timeout)
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client is not None and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client (lazy initialization)."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(timeout=self.timeout)
        return self._client

    async def invoke(self, action: str, **params: Any) -> Any:
        """Send a request to AnkiConnect API.

        Args:
            action: The AnkiConnect action name.
            **params: Parameters for the action.

        Returns:
            The result from AnkiConnect.

        Raises:
            AnkiConnectError: If the request fails or AnkiConnect returns an error.
        """
        payload: dict[str, Any] = {
            "action": action,
            "version": ANKI_CONNECT_VERSION,
        }
        if params:
            payload["params"] = params

        try:
            client = await self._get_client()
            response = await client.post(self.url, json=payload)
            response.raise_for_status()
        except httpx.ConnectError as e:
            raise AnkiConnectError(
                f"Failed to connect to AnkiConnect at {self.url}. "
                "Is Anki running with AnkiConnect installed?",
                context={"action": action},
            ) from e
        except httpx.HTTPError as e:
            raise AnkiConnectError(
                f"HTTP error communicating with AnkiConnect: {e}",
                context={"action": action},
            ) from e

        result = response.json()

        if result.get("error") is not None:
            raise AnkiConnectError(
                result["error"],
                context={"action": action},
            )

        return result.get("result")

    # -- Connection & Version --------------------------------------------------

    async def ping(self) -> bool:
        """Check if AnkiConnect is reachable."""
        try:
            version = await self.invoke("version")
            logger.debug("anki_connect_ping", version=version)
            return version is not None
        except AnkiConnectError:
            return False

    async def version(self) -> int | None:
        """Get the AnkiConnect version, or None if unreachable."""
        try:
            return cast("int", await self.invoke("version"))
        except AnkiConnectError:
            return None

    # -- Deck Operations -------------------------------------------------------

    async def deck_names(self) -> list[str]:
        """Get all deck names."""
        return cast("list[str]", await self.invoke("deckNames"))

    async def create_deck(self, deck_name: str) -> int:
        """Create a deck (no-op if it already exists). Returns deck ID."""
        return cast("int", await self.invoke("createDeck", deck=deck_name))

    async def delete_decks(
        self,
        deck_names: list[str],
        *,
        cards_too: bool = True,
    ) -> None:
        """Delete decks. cards_too must be True since Anki 2.1.28."""
        if not deck_names:
            return
        await self.invoke("deleteDecks", decks=deck_names, cardsToo=cards_too)

    # -- Note Operations -------------------------------------------------------

    async def find_notes(self, query: str) -> list[int]:
        """Find note IDs matching an Anki search query."""
        return cast("list[int]", await self.invoke("findNotes", query=query))

    async def notes_info(self, note_ids: list[int]) -> list[dict[str, Any]]:
        """Get detailed info for notes. Returns raw API dicts."""
        if not note_ids:
            return []
        return cast("list[dict[str, Any]]", await self.invoke("notesInfo", notes=note_ids))

    async def add_note(
        self,
        deck_name: str,
        model_name: str,
        fields: dict[str, str],
        tags: list[str] | None = None,
        *,
        allow_duplicate: bool = False,
    ) -> int | None:
        """Add a new note. Returns note ID, or None if duplicate."""
        note: dict[str, Any] = {
            "deckName": deck_name,
            "modelName": model_name,
            "fields": fields,
            "tags": tags or [],
            "options": {"allowDuplicate": allow_duplicate},
        }
        try:
            return cast("int", await self.invoke("addNote", note=note))
        except AnkiConnectError as e:
            if "duplicate" in str(e).lower():
                return None
            raise

    async def update_note_fields(
        self,
        note_id: int,
        fields: dict[str, str],
    ) -> None:
        """Update fields of an existing note."""
        await self.invoke(
            "updateNoteFields",
            note={"id": note_id, "fields": fields},
        )

    async def delete_notes(self, note_ids: list[int]) -> None:
        """Delete notes by ID."""
        if not note_ids:
            return
        await self.invoke("deleteNotes", notes=note_ids)

    # -- Tag Operations --------------------------------------------------------

    async def get_tags(self) -> list[str]:
        """Get all tags from the collection."""
        return cast("list[str]", await self.invoke("getTags"))

    async def add_tags(self, note_ids: list[int], tags: str) -> None:
        """Add space-separated tags to notes."""
        if not note_ids or not tags:
            return
        await self.invoke("addTags", notes=note_ids, tags=tags)

    async def remove_tags(self, note_ids: list[int], tags: str) -> None:
        """Remove space-separated tags from notes."""
        if not note_ids or not tags:
            return
        await self.invoke("removeTags", notes=note_ids, tags=tags)

    # -- Model Operations ------------------------------------------------------

    async def model_names(self) -> list[str]:
        """Get all note type (model) names."""
        return cast("list[str]", await self.invoke("modelNames"))

    async def model_field_names(self, model_name: str) -> list[str]:
        """Get field names for a note type."""
        return cast("list[str]", await self.invoke("modelFieldNames", modelName=model_name))

    # -- Sync ------------------------------------------------------------------

    async def sync(self) -> None:
        """Trigger Anki sync."""
        await self.invoke("sync")

File: crates/anki-reader/src/connect.rs
use common::error::{AnkiAtlasError, Result};
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};
use std::collections::HashMap;
use tracing::instrument;

/// Default AnkiConnect URL.
pub const ANKI_CONNECT_URL: &str = "http://localhost:8765";
/// AnkiConnect protocol version.
pub const ANKI_CONNECT_VERSION: u32 = 6;
/// Default request timeout in seconds.
pub const DEFAULT_TIMEOUT_SECS: u64 = 30;

/// Async HTTP client for the AnkiConnect API.
pub struct AnkiConnectClient {
    url: String,
    timeout: std::time::Duration,
    client: reqwest::Client,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct AnkiConnectNoteField {
    pub value: String,
    pub order: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct AnkiConnectNoteInfo {
    #[serde(rename = "noteId")]
    pub note_id: i64,
    #[serde(rename = "modelName")]
    pub model_name: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub fields: HashMap<String, AnkiConnectNoteField>,
    #[serde(default)]
    pub cards: Vec<i64>,
}

/// Controls duplicate note handling.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DuplicateHandling {
    Allow,
    Reject,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AddNoteOutcome {
    Added(i64),
    Duplicate,
}

#[derive(Serialize)]
struct AnkiConnectRequest<'a> {
    action: &'a str,
    version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    params: Option<Value>,
}

impl AnkiConnectClient {
    pub fn new(url: &str, timeout_secs: u64) -> Self {
        Self {
            url: url.to_string(),
            timeout: std::time::Duration::from_secs(timeout_secs),
            client: reqwest::Client::new(),
        }
    }

    fn action_error(&self, action: &str, message: impl Into<String>) -> AnkiAtlasError {
        AnkiAtlasError::AnkiConnect {
            message: message.into(),
            context: HashMap::from([("action".into(), action.to_string())]),
        }
    }

    async fn invoke_value(&self, action: &str, params: Option<Value>) -> Result<Value> {
        let payload = AnkiConnectRequest {
            action,
            version: ANKI_CONNECT_VERSION,
            params,
        };

        let response = self
            .client
            .post(&self.url)
            .json(&payload)
            .timeout(self.timeout)
            .send()
            .await
            .map_err(|e| {
                let message = if e.is_connect() {
                    format!(
                        "AnkiConnect not reachable at {}. Is Anki running?",
                        self.url
                    )
                } else {
                    format!("HTTP error: {e}")
                };
                self.action_error(action, message)
            })?;

        let body: Value = response
            .json()
            .await
            .map_err(|e| self.action_error(action, format!("failed to parse response: {e}")))?;

        if let Some(error) = body.get("error").and_then(|e| e.as_str()) {
            return Err(self.action_error(action, error));
        }

        body.get("result")
            .cloned()
            .ok_or_else(|| self.action_error(action, "response missing result field"))
    }

    async fn invoke_typed<T: DeserializeOwned>(
        &self,
        action: &str,
        params: Option<Value>,
    ) -> Result<T> {
        let result = self.invoke_value(action, params).await?;
        serde_json::from_value(result)
            .map_err(|e| self.action_error(action, format!("failed to parse result: {e}")))
    }

    async fn invoke_unit(&self, action: &str, params: Option<Value>) -> Result<()> {
        let result = self.invoke_value(action, params).await?;
        if result.is_null() {
            Ok(())
        } else {
            Err(self.action_error(action, "expected null result"))
        }
    }

    #[instrument(skip(self))]
    pub async fn ping(&self) -> Result<()> {
        self.version().await.map(|_| ())
    }

    #[instrument(skip(self))]
    pub async fn version(&self) -> Result<u32> {
        self.invoke_typed("version", None).await
    }

    #[instrument(skip(self))]
    pub async fn deck_names(&self) -> Result<Vec<String>> {
        self.invoke_typed("deckNames", None).await
    }

    #[instrument(skip(self))]
    pub async fn create_deck(&self, name: &str) -> Result<i64> {
        self.invoke_typed("createDeck", Some(json!({"deck": name})))
            .await
    }

    #[instrument(skip(self))]
    pub async fn delete_decks(&self, names: &[String], cards_too: bool) -> Result<()> {
        if names.is_empty() {
            return Ok(());
        }
        self.invoke_unit(
            "deleteDecks",
            Some(json!({"decks": names, "cardsToo": cards_too})),
        )
        .await?;
        Ok(())
    }

    #[instrument(skip(self))]
    pub async fn find_notes(&self, query: &str) -> Result<Vec<i64>> {
        self.invoke_typed("findNotes", Some(json!({"query": query})))
            .await
    }

    #[instrument(skip(self))]
    pub async fn notes_info(&self, note_ids: &[i64]) -> Result<Vec<AnkiConnectNoteInfo>> {
        if note_ids.is_empty() {
            return Ok(Vec::new());
        }
        self.invoke_typed("notesInfo", Some(json!({"notes": note_ids})))
            .await
    }

    #[instrument(skip(self, fields, tags))]
    pub async fn add_note(
        &self,
        deck_name: &str,
        model_name: &str,
        fields: &HashMap<String, String>,
        tags: &[String],
        allow_duplicate: DuplicateHandling,
    ) -> Result<AddNoteOutcome> {
        let mut note = json!({
            "deckName": deck_name,
            "modelName": model_name,
            "fields": fields,
            "tags": tags,
        });

        if allow_duplicate == DuplicateHandling::Allow {
            note["options"] = json!({"allowDuplicate": true});
        }

        match self
            .invoke_typed("addNote", Some(json!({"note": note})))
            .await
        {
            Ok(note_id) => Ok(AddNoteOutcome::Added(note_id)),
            Err(AnkiAtlasError::AnkiConnect { ref message, .. })
                if message.to_lowercase().contains("duplicate") =>
            {
                Ok(AddNoteOutcome::Duplicate)
            }
            Err(e) => Err(e),
        }
    }

    #[instrument(skip(self, fields))]
    pub async fn update_note_fields(
        &self,
        note_id: i64,
        fields: &HashMap<String, String>,
    ) -> Result<()> {
        self.invoke_unit(
            "updateNoteFields",
            Some(json!({"note": {"id": note_id, "fields": fields}})),
        )
        .await?;
        Ok(())
    }

    #[instrument(skip(self))]
    pub async fn delete_notes(&self, note_ids: &[i64]) -> Result<()> {
        if note_ids.is_empty() {
            return Ok(());
        }
        self.invoke_unit("deleteNotes", Some(json!({"notes": note_ids})))
            .await?;
        Ok(())
    }

    #[instrument(skip(self))]
    pub async fn get_tags(&self) -> Result<Vec<String>> {
        self.invoke_typed("getTags", None).await
    }

    #[instrument(skip(self, note_ids))]
    pub async fn add_tags(&self, note_ids: &[i64], tags: &str) -> Result<()> {
        if note_ids.is_empty() || tags.is_empty() {
            return Ok(());
        }
        self.invoke_unit("addTags", Some(json!({"notes": note_ids, "tags": tags})))
            .await?;
        Ok(())
    }

    #[instrument(skip(self, note_ids))]
    pub async fn remove_tags(&self, note_ids: &[i64], tags: &str) -> Result<()> {
        if note_ids.is_empty() || tags.is_empty() {
            return Ok(());
        }
        self.invoke_unit("removeTags", Some(json!({"notes": note_ids, "tags": tags})))
            .await?;
        Ok(())
    }

    #[instrument(skip(self))]
    pub async fn model_names(&self) -> Result<Vec<String>> {
        self.invoke_typed("modelNames", None).await
    }

    #[instrument(skip(self))]
    pub async fn model_field_names(&self, model_name: &str) -> Result<Vec<String>> {
        self.invoke_typed("modelFieldNames", Some(json!({"modelName": model_name})))
            .await
    }

    #[instrument(skip(self))]
    pub async fn sync(&self) -> Result<()> {
        self.invoke_unit("sync", None).await?;
        Ok(())
    }
}

impl Default for AnkiConnectClient {
    fn default() -> Self {
        Self::new(ANKI_CONNECT_URL, DEFAULT_TIMEOUT_SECS)
    }
}
